### PR TITLE
Parquet Writer: Early out creating dictionary

### PR DESCRIPTION
**Pull request by lnkuiper**
We currently create the whole dictionary, and only after creating it do we check if the compression ratio is greater than 1. If yes, we apply dictionary compression.

However, creating the dictionary requires a lot of string hashing and comparisons. This PR checks whether the compression ratio is less than 1 after inserting 10k values into the dictionary. If yes, we stop inserting into the dictionary, preventing costly string hashing and comparisons. This speeds up writing lineitem at SF10 by 15-20%, as it can early out for the `l_comment` column.

Not sure if this is the best fix, so I'm happy to receive any feedback. I've sent this PR to `main`, but I'm also happy to change it to `feature` instead.

### CODE REVIEW

**Review by Mytherin — COMMENTED** [review]
Thanks! I think this is a good idea, but I would like for it to be configurable through a setting. That way we can also disable dictionary writing entirely using the same setting by setting the ratio to 0 (which has been requested before). Could we also add some tests that trigger explicitly this behavior?

**Comment by lnkuiper** [comment]
@Mytherin, thanks for the feedback. I've now made it configurable through a parameter `dictionary_compression_ratio_threshold`. When set to 0, dictionary compression is always applied, when set to -1, dictionary compression is never applied.

I thought it made more sense this way, i.e., if you only want to use dictionary compression when you have a good ratio, you can set the parameter to e.g., 2.0, and it will start applying dictionary compression when the compression ratio is greater than 2.0.

File: test/sql/copy/parquet/dictionary_compression_ratio_threshold.test [hunk 1]
```diff
@@ -0,0 +1,80 @@
+# name: test/sql/copy/parquet/dictionary_compression_ratio_threshold.test
+# description: Test Parquet dictionary_compression_ratio_threshold parameter
+# group: [parquet]
+
+require parquet
+
+statement ok
+CREATE TABLE test AS SELECT 'thisisaverylongstringbutitrepeatsmanytimessoitshighlycompressible' || (range % 10) i FROM range(100000)
+
+mode skip
```
Comment (Mytherin) [inline test/sql/copy/parquet/dictionary_compression_ratio_threshold.test:10]: Is this left in on purpose?

Comment (lnkuiper) [reply test/sql/copy/parquet/dictionary_compression_ratio_threshold.test]: No! I've removed the skip now

**Comment by lnkuiper** [comment]
> Should there be a very early out if the ratio is impossible to achieve that skips the routine at all?

If you set the parameter to `-1` you get out immediately, and waste no time at all.

> And the check on compression better than 1.0 at the end should be restored? (might also make sense to say that values between 0 and 1.0 are basically useful only for testing)

Why? Then the supplied parameter wouldn't be respected no?

**Comment by carlopi** [comment]
> If you set the parameter to `-1` you get out immediately, and waste no time at all.
+1

> Why? Then the supplied parameter wouldn't be respected no?
Unsure, maybe elsewhere we are relying on some bound on size, unsure.

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**